File: src/main/java/com/handwoong/everyonewaiter/domain/Member.java
package com.handwoong.everyonewaiter.domain;

import static lombok.AccessLevel.PRIVATE;
import static lombok.AccessLevel.PROTECTED;

import com.handwoong.everyonewaiter.dto.member.MemberRegisterDto;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = PROTECTED)
public class Member extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "member_id")
    private Long id;

    @NotNull
    @Column(unique = true)
    private String username;

    @NotNull
    private String password;

    @NotNull
    @Column(unique = true)
    private String phoneNumber;

    private int balance;

    @NotNull
    @Enumerated(value = EnumType.STRING)
    private MemberRole role;

    @Builder(access = PRIVATE)
    private Member(String username, String password, String phoneNumber, int balance,
            MemberRole role) {
        this.username = username;
        this.password = password;
        this.phoneNumber = phoneNumber;
        this.balance = balance;
        this.role = role;
    }

    public static Member createMember(MemberRegisterDto memberDto) {
        return Member.builder()
                .username(memberDto.getUsername())
                .password(memberDto.getPassword())
                .phoneNumber(memberDto.getPhoneNumber())
                .balance(0)
                .role(MemberRole.USER)
                .build();
    }

    public void encodePassword(String encodePassword) {
        this.password = encodePassword;
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/domain/member/Member.kt
package com.handwoong.everyonewaiter.domain.member

import com.handwoong.everyonewaiter.domain.BaseEntity
import com.handwoong.everyonewaiter.domain.member.MemberRole.ROLE_USER
import com.handwoong.everyonewaiter.dto.member.MemberCreateRequest
import javax.persistence.*
import javax.persistence.GenerationType.IDENTITY
import javax.validation.constraints.NotNull

@Entity
class Member(

    @NotNull
    @Column(unique = true)
    val username: String,

    @NotNull
    var password: String,

    @NotNull
    @Column(unique = true)
    val phoneNumber: String,

    var balance: Int = 300,

    @NotNull
    @Enumerated(value = EnumType.STRING)
    var role: MemberRole = ROLE_USER,

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "member_id")
    val id: Long? = null,
) : BaseEntity() {

    fun encodePassword(encodePassword: String) {
        this.password = encodePassword
    }

    companion object {
        fun createMember(memberDto: MemberCreateRequest): Member {
            return Member(
                username = memberDto.username.lowercase(),
                password = memberDto.password,
                phoneNumber = memberDto.phoneNumber,
            )
        }
    }

}
